### PR TITLE
Table View Edit Columns Dropdown

### PR DESCRIPTION
**Pull request by TanayParikh**
- Adds support for editing columns displayed in a table view. (Vertical ellipses after last column header). 

## Components Implemented
- Overlay Dropdown (can also be used for an account/login menu for example)
- Edit Columns View
- Updated Checkbox to support labels

**TODO: Shift overlay dropdown on screen resize.**

### CODE REVIEW

File: src/com/google/foam/demos/u2/Demo.js [hunk 1]
```diff
@@ -213,6 +213,92 @@ var i = E('image').attrs({
 i.write();
 
 foam.CLASS({
+  name: 'Linking',
```
Comment (kgrgreer) [inline src/com/google/foam/demos/u2/Demo.js]: Don't submit this, it was just for your debugging purposes.

File: src/foam/u2/md/OverlayDropdown.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+      class: 'Float',
+      name: 'height',
+      value: 0
+    },
+    {
+      class: 'Boolean',
+      name: 'opened',
+      documentation: 'True when the overlay has been commanded to be open. ' +
+          'It might still be animating; see $$DOC{ref:".animationComplete"}.',
+      value: false
```
Comment (kgrgreer) [inline src/foam/u2/md/OverlayDropdown.js]: False is the default default value so this line isn't needed.

File: src/foam/u2/md/OverlayDropdown.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+      name: 'animationComplete',
+      documentation: 'True when an animation is running. The overlay hasn\'t ' +
+          'really reached the state commanded by $$DOC{ref:".opened"} until ' +
+          'this is true.',
+      value: true
+    },
+    {
+      name: 'dropdownE_',
+      factory: function() {
+        // This needs to be created early so it's safe to add() things to it.
```
Comment (kgrgreer) [inline src/foam/u2/md/OverlayDropdown.js]: I don't think this is true.

File: src/com/google/foam/demos/u2/Demo.js (hunk 1 above)
Comment (TanayParikh) [reply src/com/google/foam/demos/u2/Demo.js]: Done.

File: src/foam/u2/md/OverlayDropdown.js [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    },
+    {
+      name: 'addToSelf_',
+      value: false
+    }
+  ],
+
+  methods: [
+    function add() {
+      if (this.addToSelf_) this.SUPER.apply(this, arguments);
```
Comment (kgrgreer) [inline src/foam/u2/md/OverlayDropdown.js]: Follow the foam style guide: if ( this.addToSelf_ ) (extra spaces).
Don't form if/else's this way.

File: src/foam/u2/md/OverlayDropdown.js (hunk 2 above)
Comment (TanayParikh) [reply src/foam/u2/md/OverlayDropdown.js]: Done.

File: src/foam/u2/md/OverlayDropdown.js [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+      this.addToSelf_ = true;
+      this.addClass(this.myClass('container'));
+      var view = this;
+
+      this.addClass(this.slot(function(open) {
+        this.shown = open;
+      }, this.opened$));
+
+      this.start('dropdown-overlay')
+          .addClass(this.myClass('overlay'))
```
Comment (kgrgreer) [inline src/foam/u2/md/OverlayDropdown.js]: Indent 2 spaces. Have end() line up with this.

File: src/foam/u2/view/EditColumnsView.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+      //   var props = this.properties;
+      //   props = this.properties.slice();
+      //   props.sort(function(a, b) {
+      //     return a.label.toLowerCase().compareTo(b.label.toLowerCase());
+      //   });
+      // } else { ...
+
+      this.selected = []
+
+      for (let i = 0; i < this.columns_.length; i++) {
```
Comment (kgrgreer) [inline src/foam/u2/view/EditColumnsView.js]: Spacing.

File: src/foam/u2/view/EditColumnsView.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        this.selected.push(cb.data$);
+        var name = this.columns_[i].name;
+
+        // Subscribes updateTable listener to checkbox data
+        cb.data$.sub(this.updateTable.bind(this, name));
+
+        this.add(cb);
+
+        // Ensures each selection is on a new line
+        if (i != this.columns_.length - 1) this.start('br').end();
```
Comment (kgrgreer) [inline src/foam/u2/view/EditColumnsView.js]: Spacing here and elsewhere.

File: src/foam/u2/view/TableView.js [hunk 8]
```diff
@@ -155,6 +175,33 @@ foam.CLASS({
             map(foam.core.Property.NAME.f);
       }
     },
+    {
+      class: 'Boolean',
+      name: 'editColumnsEnabled',
+      documentation: 'Set this to true to let the user select columns.',
+      value: false
```
Comment (kgrgreer) [inline src/foam/u2/view/TableView.js]: Default default.

File: src/foam/u2/view/TableView.js [hunk 9]
```diff
@@ -166,8 +213,39 @@ foam.CLASS({
...
+      return this.OverlayDropdown.create().add(editor);
+    },
+
+    /** Adds offset for edit columns overlay dropdown
+     * OverlayDropdown adds element to top right of parent container.
+     * We want the table dropdown to appear below the dropdown icon.
+     */
+    function positionOverlayDropdown(columnSelectionE) {
+      // Dynamic position calculation
+      var origin = document.getElementsByClassName(this.of.id.replace(/\./g,'-')
```
Comment (kgrgreer) [inline src/foam/u2/view/TableView.js]: All wrong.

File: src/foam/u2/view/TableView.js [hunk 10]
```diff
@@ -182,12 +260,27 @@ foam.CLASS({
...
+                    addClass(view.myClass('th-editColumns')).
+                      on('click', function(e) {
+                        view.positionOverlayDropdown(columnSelectionE);
+                        columnSelectionE.open();
+                      }).
+                      add(' ', view.vertMenuIcon).
+                        addClass(view.myClass('vertDots')).
+                        addClass(view.myClass('noselect')).
+                        start('div').
+                        addClass(view.of.id.replace(/\./g,'-') +
```
Comment (kgrgreer) [inline src/foam/u2/view/TableView.js]: Wrong.

File: src/foam/u2/view/EditColumnsView.js (hunk 7 above)
Comment (TanayParikh) [reply src/foam/u2/view/EditColumnsView.js]: Fixed.

File: src/foam/u2/view/EditColumnsView.js (hunk 6 above)
Comment (TanayParikh) [reply src/foam/u2/view/EditColumnsView.js]: Fixed.

File: src/foam/u2/view/TableView.js (hunk 8 above)
Comment (TanayParikh) [reply src/foam/u2/view/TableView.js]: Removed.

File: src/foam/u2/view/TableView.js (hunk 9 above)
Comment (TanayParikh) [reply src/foam/u2/view/TableView.js]: Fixed.

File: src/foam/u2/view/TableView.js (hunk 10 above)
Comment (TanayParikh) [reply src/foam/u2/view/TableView.js]: Updated.

File: src/foam/u2/md/OverlayDropdown.js (hunk 4 above)
Comment (TanayParikh) [reply src/foam/u2/md/OverlayDropdown.js]: Updated spacing, didn't see how `else` is to be handled: https://google.github.io/styleguide/javascriptguide.xml

File: src/foam/u2/md/OverlayDropdown.js (hunk 5 above)
Comment (TanayParikh) [reply src/foam/u2/md/OverlayDropdown.js]: Updated.

File: src/foam/u2/view/EditColumnsView.js (hunk 6 above)
Comment (kirkeaton) [reply src/foam/u2/view/EditColumnsView.js]: @kgrgreer Shouldn't this also be `var i = 0;` instead of `let i = 0;`?

Comment (TanayParikh) [reply src/foam/u2/view/EditColumnsView.js]: Updated. Fixed spacing as well.

File: src/foam/u2/CheckBox.js [hunk 11]
```diff
@@ -23,16 +23,68 @@ foam.CLASS({
...
+      name: 'label' 
+    }
   ],
 
   methods: [
     function initE() {
       this.SUPER();
       this.setAttribute('type', 'checkbox');
+
+      if (this.showLabel) {
```
Comment (kgrgreer) [inline src/foam/u2/CheckBox.js]: See style guide, should be:
if ( this.showLabel ) {
2 space indents.

File: src/foam/u2/md/OverlayDropdown.js [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+    },
+    {
+      name: 'addToSelf_',
+      value: false
+    }
+  ],
+
+  methods: [
+    function add() {
+      if ( this.addToSelf_ ) this.SUPER.apply(this, arguments);
```
Comment (kgrgreer) [inline src/foam/u2/md/OverlayDropdown.js]: Don't format if/else this way
if
  something
else
  something else

File: src/foam/u2/md/OverlayDropdown.js [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+    {
+      class: 'Boolean',
+      name: 'animationComplete',
+      documentation: 'True when an animation is running. The overlay hasn\'t ' +
+          'really reached the state commanded by $$DOC{ref:".opened"} until ' +
+          'this is true.',
+      value: true
+    },
+    {
+      name: 'dropdownE_',
```
Comment (kgrgreer) [inline src/foam/u2/md/OverlayDropdown.js:100]: Why is this needed?

File: src/foam/u2/md/OverlayDropdown.js [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+          .on('mouseleave', this.onMouseLeave)
+          .on('click', this.onClick);
+
+      this.add(this.dropdownE_);
+
+      this.addToSelf_ = false;
+    }
+  ],
+
+  axioms: [
```
Comment (kgrgreer) [inline src/foam/u2/md/OverlayDropdown.js]: Convention is to put axioms before properties.

File: src/foam/u2/view/EditColumnsView.js [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.u2.view',
+  name: 'EditColumnsView',
+  extends: 'foam.u2.Element',
+
+  requires: [
+      'foam.u2.CheckBox'
```
Comment (kgrgreer) [inline src/foam/u2/view/EditColumnsView.js]: Two space indent here and below.

File: src/foam/u2/md/OverlayDropdown.js (hunk 13 above)
Comment (TanayParikh) [reply src/foam/u2/md/OverlayDropdown.js:100]: It is the actual dropdown element being worked with.

File: src/foam/u2/md/OverlayDropdown.js (hunk 12 above)
Comment (TanayParikh) [reply src/foam/u2/md/OverlayDropdown.js]: Thanks, updated to:

```js
      if ( this.addToSelf_ ) {
        this.SUPER.apply(this, arguments);
      } else {
        this.dropdownE_.add.apply(this.dropdownE_, arguments);
      }
```

File: src/foam/u2/md/OverlayDropdown.js (hunk 14 above)
Comment (TanayParikh) [reply src/foam/u2/md/OverlayDropdown.js]: Moved, thanks.

File: src/foam/u2/CheckBox.js (hunk 11 above)
Comment (TanayParikh) [reply src/foam/u2/CheckBox.js]: Updated.

File: src/foam/u2/md/OverlayDropdown.js (hunk 3 above)
Comment (TanayParikh) [reply src/foam/u2/md/OverlayDropdown.js]: Removed.

File: src/foam/u2/view/EditColumnsView.js (hunk 15 above)
Comment (TanayParikh) [reply src/foam/u2/view/EditColumnsView.js]: Updated, thanks.

File: src/foam/u2/view/EditColumnsView.js [hunk 16]
```diff
@@ -46,7 +46,7 @@ foam.CLASS({
 
       this.selected = []
 
-      for (let i = 0; i < this.columns_.length; i++) {
+      for ( var i = 0; i < this.columns_.length; i++ ) {
```
Comment (kgrgreer) [inline src/foam/u2/view/EditColumnsView.js]: Also, spaces before the ;'s.

File: src/foam/u2/view/EditColumnsView.js [hunk 17]
```diff
@@ -82,7 +82,7 @@ foam.CLASS({
       // }
       // else { ...
 
-      for (var i = 0; i < this.columns.length; i++) {
+      for ( var i = 0; i < this.columns.length; i++ ) {
```
Comment (kgrgreer) [inline src/foam/u2/view/EditColumnsView.js]: Also, spaces before the ;'s.

Comment (TanayParikh) [reply src/foam/u2/view/EditColumnsView.js]: Added.

File: src/foam/u2/view/EditColumnsView.js (hunk 16 above)
Comment (TanayParikh) [reply src/foam/u2/view/EditColumnsView.js]: Added.

**Comment by TanayParikh** [comment]
Also updated spacing for the `Enum.js` Java generation class.

File: src/foam/u2/CheckBox.js [hunk 18]
```diff
@@ -23,16 +23,68 @@ foam.CLASS({
...
+
+      if ( this.showLabel ) { 
+        this.start('label')
+          .addClass(this.myClass('label'))
+          .addClass(this.myClass('noselect'))
+          .add(this.label$)
+          .on('click', function() { 
+            this.data = !this.data; 
+          }.bind(this))
+          .end();
```
Comment (kgrgreer) [inline src/foam/u2/CheckBox.js]: Line .end() up with this.start().

Comment (TanayParikh) [reply src/foam/u2/CheckBox.js]: Done.

File: src/foam/u2/md/OverlayDropdown.js (hunk 13 above)
Comment (kgrgreer) [reply src/foam/u2/md/OverlayDropdown.js:100]: Do you need to build this in the factory (which means it might not get the proper context)? Can't you just build it like normal and then just specify dropdownE_$ as the third argument to start() or tag() to store the reference here?

File: src/foam/u2/view/TableView.js [hunk 19]
```diff
@@ -182,12 +261,25 @@ foam.CLASS({
...
+                  return this.E('th').
+                    addClass(view.myClass('th-editColumns')).
+                      on('click', function(e) {
+                        view.positionOverlayDropdown(columnSelectionE);
+                        columnSelectionE.open();
+                      }).
+                      add(' ', view.vertMenuIcon).
+                        addClass(view.myClass('vertDots')).
+                        addClass(view.myClass('noselect')).
+                        start('div', null, this.dropdownOrigin$).end();
```
Comment (kgrgreer) [inline src/foam/u2/view/TableView.js]: start().end() is just the same as tag().

File: src/foam/u2/view/TableView.js [hunk 20]
```diff
@@ -214,7 +306,10 @@ foam.CLASS({
                             column.f ? column.f(obj) : null, obj, column
                           ]).
                           end();
-                      });
+                      }).
+                      add(this.slot(function(editColumnsEnabled) {
```
Comment (kgrgreer) [inline src/foam/u2/view/TableView.js]: Does this need to be a slot? Are we going to change this setting on an already created table?

File: src/foam/u2/view/TableView.js (hunk 19 above)
Comment (TanayParikh) [reply src/foam/u2/view/TableView.js]: Updated.

File: src/foam/u2/view/TableView.js (hunk 20 above)
Comment (TanayParikh) [reply src/foam/u2/view/TableView.js]: Replaced with `call()`

File: src/foam/u2/md/OverlayDropdown.js (hunk 13 above)
Comment (TanayParikh) [reply src/foam/u2/md/OverlayDropdown.js:100]: Added TODO 25c233e
